fix(query-source): Fix query source relative filepath

gggritso opened the pull request:
We're seeing an issue in Sentry where query source file paths aren't matched to GitHub URLs because the path prefixes are different from the prefixes generated for stack traces for the same files.

e.g., in `getsentry/sentry` good path is like `sentry/api/serializers/base.py`, but query sources for that file have the file path `sentry/src/sentry/api/serializers/base.py`:  we're adding an extra `sentry/src` prefix to the path. The existing GitHub file path mappings don't work, and there's no "Open in GitHub" link as a result.

The reason is that for stack traces, relative file paths are made using the `filename_for_module` function. It checks currently loaded modules, and removes them from the path prefix. For query sources, the logic is simplier, it's `filepath.replace(project_root, "").lstrip(os.sep)`.

This PR changes query source paths to use `filename_for_module` just like stack frame paths. In most cases this won't make a difference, but for Sentry's setup it should make sure that relative file paths are consistent between stack traces and query sources for a given file.

sentrivana commented:
Heyy @gggritso 👋🏻 The change and the reasoning make sense to me.

As for testing this: we usually try to make sure each change is covered by a test. In this case there are already testcases for query source in the three different integrations where we support it, so ideally this should just be a matter of taking an existing testcase, copying it and modifying it slightly to trigger behavior similar to what you're seeing in Sentry.

I think these specific test cases could be repurposed easily:
* https://github.com/getsentry/sentry-python/blob/c77a1235f4d4f4d88129c13fa9586840ede48ce4/tests/integrations/django/test_db_query_data.py#L120 for Django
* https://github.com/getsentry/sentry-python/blob/c77a1235f4d4f4d88129c13fa9586840ede48ce4/tests/integrations/sqlalchemy/test_sqlalchemy.py#L392 for SQLAlchemy
* https://github.com/getsentry/sentry-python/blob/c77a1235f4d4f4d88129c13fa9586840ede48ce4/tests/integrations/asyncpg/test_asyncpg.py#L549 for asyncpg

You'll probably need to add some custom `project_root` to the `sentry_init` and then check that `data.get(SPANDATA.CODE_FILEPATH)` looks the way it should.

As for Python 2 support, at this stage we still support it 100%, so ideally after this change you'd get the same filepath regardless of Python version. Verifying this locally is a bit tricky because you need some `pyenv`-like multiple py version setup, but the CI will run all tests suites in Python 2 as well, so as long as you write the new tests cases and they pass the CI, that should be enough to verify that it works in py2 too.

gggritso commented:
@sentrivana thanks for the hints! I added a unit test for each of the three integrations you mentioned. It was a little complicated, so I updated the PR description to explain everything that's going on (it wasn't a `project_root` thing, it was more of a module import thing).

As for Python 2, I'm not sure I can get it to work perfectly. More in the PR, but the behaviour there is different enough that it doesn't really work. You tell me, but since this is an _improvement_ to a feature (Python 2 still has a good graceful fallback!) and I'm not breaking anything existing, it might be worth just merging.

P.S. Do you know how long Sentry will keep supporting Python 2?

sentrivana commented on tests/integrations/asyncpg/__init__.py:
```diff
@@ -1,4 +1,8 @@
 pytest.importorskip("asyncpg")
 pytest.importorskip("pytest_asyncio")
+
+sys.path.insert(0, os.path.join(os.path.dirname(__file__)))
```
Can we add a comment here explaining why this is here?

sentrivana commented on tests/integrations/asyncpg/asyncpg_helpers/helpers.py:
Nit: I'd just call the directory `helpers` since it's already in an `asyncpg` dir/namespace.

It'd also be cool to have a comment either here or where the function is used explaining why this exists and why it needs to be in a separate directory.

sentrivana reviewed (approved):
LGTM, left two nits on the asyncpg tests but they apply to all three test suites.

gggritso commented:
@sentrivana I added some comments to explain why the `sys.path` editing is there, let me know if you want me to add more, plus a docstring to the tests

As for the dir/namespace, I'm a little stuck there. I'd _like_ to rename `sqlalchemy_helpers` to `helpers`, but if the helper directories in `asyncpg`, `sqlalchemy` and `django` are all called `helpers` there will be a conflict at import time, no?

I thought about it for a while, and tried to re-do the tests so they only edit `sys.path` for the one test, but I couldn't figure out how to do it for Django. There, the app is created as a fixture, so the routes are only imported once. I need a way to change the module search path, import a different URL config, create an app with this different config, run tests, and then clean up the module search path. I couldn't figure out a clean way of doing it. What do you think?

sentrivana commented:
@gggritso I see -- I think in that case it's ok to leave the `*_helpers` dirs as is. There's probably a way around it but I also don't have any ideas atm. Feel free to merge :shipit: (You might need to click the `Update branch` button before you can do that though.)
